clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -601,7 +603,9 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                     error);
                 if (transactionManager == null) {
                     reenqueueBatch(batch, now);
-                } else if (transactionManager.hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch())) {
+                } else if (transactionManager.hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch()) ||
+                        (transactionManager.hasProducerIdAndEpoch(batch.producerId(), (short) (batch.producerEpoch() + 1)) &&
```

Comment:
If we are trying to handle epoch bump case, do we also need to handle the case when we got a new producerId after exhausting epochs?

Reply:
We would have, yes, but I moved this check inside TransactionManager before we do the bump (which is safe because we modify all the in-flights at the time of the bump).